### PR TITLE
Replace fetching updates from updates.mautic.org to use Github releases

### PR DESCRIPTION
**Pull request by alanhartless**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

Similar to https://github.com/mautic/mautic/pull/8226, this PR is removing another dependency on https://updates.mautic.org by pulling releases directly from Github. It uses Github's public releases API endpoint to get a list of releases. It then checks for a `app/release_metadata.json` file that contains min/max supported PHP version, stability, announcement URL (will use Github's release if empty), and minimum supported Mautic version. It then determines the applicable latest supported release. 

(Note - download statistics can be collected using Github's release API for the assets)

### 
[//]: # ( As applicable: )

#### Steps to test this PR:
1. Install this PR as a new Mautic installation
2. add `    'system_update_url'         => 'https://api.github.com/repos/alanhartless/mautic/releases',` to `app/config/local.php` and clear the cache. JUST KIDDING, you don't have the clear the cache any more for most local.php changes. 🎉
3. Login and note that bottom right has the version as 3.0.0-beta
4. There should be a notification for the 3.0.0 update being available
5. Go through the in-app update
6. Mautic should now show 3.0.0 as the installed version
7.  Repeat with a new installation and this time use the update command `php bin/console mautic:update:find` then `php bin/console mautic:update:apply` and then finally `php bin/console mautic:update:apply --finish`.

#### List backwards compatibility breaks:
1. `\AppKernel::MAJOR_VERSION`, `\AppKernel::MINOR_VERSION`, `\AppKernel::PATCH_VERSION`, and `\AppKernel::EXTRA_VERSION` are no longer defined. Use `\Mautic\CoreBundle\Release\ThisRelease::getMetadata()->getMajorVersion()`, etc instead. 
2. `\Mautic\CoreBundle\Helper\UpdateHelper::getServerOs` is now private

### CODE REVIEW

File: app/bundles/CoreBundle/Helper/UpdateHelper.php [hunk 1]
```diff
@@ -11,19 +11,21 @@
 
 namespace Mautic\CoreBundle\Helper;
 
-use Joomla\Http\Http;
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\GuzzleException;
```
Comment (hluchas) [inline app/bundles/CoreBundle/Helper/UpdateHelper.php]: This class is not used here
```suggestion
```

**Comment by lukassykora** [comment]
I got an error:  Class 'Symfony\Bundle\WebProfilerBundle\WebProfilerBundle' not found

I followed the steps:
1. Install this PR as a new Mautic installation
2. add 'system_update_url' => 'https://api.github.com/repos/alanhartless/mautic/releases', to app/config/local.php
3. Login and note that bottom right has the version as 3.0.0-beta **(it works)**
4. There should be a notification for the 3.0.0 update being available **(it works)**
5. Go through the in-app update (it does some steps and then error)

Maybe the problem is there just because of I used /index_dev.php

**Review by hluchas — APPROVED** [review]
Looks good and works great 👍 
Thank you!

**Review by hluchas — CHANGES_REQUESTED** [review]
I didn't test cli command. Here is the result.

<details><summary>With applied https://github.com/mautic/mautic/pull/8466</summary>

```bash
~/dev/test $ cat app/config/local.php
<?php
$parameters = array(
	'db_driver' => 'pdo_mysql',
	'db_host' => 'localhost',
	'db_table_prefix' => 'mautic_',
	'db_port' => '3306',
	'db_name' => 'test',
	'db_user' => 'root',
	'db_password' => null,
	'db_backup_tables' => 0,
	'db_backup_prefix' => 'bak_',
	'db_server_version' => '5.7.29',
	'mailer_from_name' => 'Lukas Drahy',
	'mailer_from_email' => 'lukas@drahy.net',
	'mailer_transport' => 'smtp',
	'mailer_host' => 'localhost',
	'mailer_port' => '1025',
	'mailer_user' => null,
	'mailer_password' => null,
	'mailer_api_key' => null,
	'mailer_encryption' => null,
	'mailer_auth_mode' => null,
	'mailer_spool_type' => 'memory',
	'mailer_spool_path' => '%kernel.root_dir%/../var/spool',
	'secret_key' => '4d315cca4fd83408a93bf1424c82911ae3fa7dffd45f876ae85eb8be9e11c9a6',
	'site_url' => 'https://test.local',
	'system_update_url' => 'https://api.github.com/repos/alanhartless/mautic/releases',
);
~/dev/test $ rm -rf var/cache/* && rm -rf var/logs/* && bin/console cache:warmup

 // Warming up the cache for the prod environment with debug false


 [OK] Cache for the "prod" environment (debug=false) was successfully warmed.


~/dev/test $ bin/console mautic:update:find
Version 3.0.0 of Mautic is available for download. Please visit https://github.com/alanhartless/mautic/releases/tag/3.0.0 for more information.
To update, you can run 'php bin/console mautic:update:apply' from the command line.
~/dev/test $ bin/console mautic:update:apply
Are you sure you wish to update Mautic to the latest version? y
Step    5 [----->----------------------] Clearing the cache                  PHP Warning:  require(/Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/getMautic_Helper_LanguageService.php): failed to open stream: No such file or directory in /Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/appProdProjectContainer.php on line 5707
PHP Stack trace:
PHP   1. {main}() /Users/lukas.drahy/dev/test/bin/console:0
PHP   2. Symfony\Bundle\FrameworkBundle\Console\Application->run() /Users/lukas.drahy/dev/test/bin/console:41
PHP   3. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:148
PHP   4. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:74
PHP   5. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:255
PHP   6. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:86
PHP   7. Mautic\CoreBundle\Command\ApplyUpdatesCommand->run() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:1001
PHP   8. Mautic\CoreBundle\Command\ApplyUpdatesCommand->execute() /Users/lukas.drahy/dev/test/vendor/symfony/console/Command/Command.php:255
PHP   9. Container9uztwgj\appProdProjectContainer->get() /Users/lukas.drahy/dev/test/app/bundles/CoreBundle/Command/ApplyUpdatesCommand.php:246
PHP  10. Container9uztwgj\appProdProjectContainer->load() /Users/lukas.drahy/dev/test/vendor/symfony/dependency-injection/Container.php:304

Warning: require(/Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/getMautic_Helper_LanguageService.php): failed to open stream: No such file or directory in /Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/appProdProjectContainer.php on line 5707

Call Stack:
    0.0003     403184   1. {main}() /Users/lukas.drahy/dev/test/bin/console:0
    0.0283    4286320   2. Symfony\Bundle\FrameworkBundle\Console\Application->run() /Users/lukas.drahy/dev/test/bin/console:41
    0.0521    4672824   3. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:148
    0.3098   36609504   4. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:74
    0.3101   36610120   5. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:255
    0.3101   36610120   6. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:86
    0.3111   36674064   7. Mautic\CoreBundle\Command\ApplyUpdatesCommand->run() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:1001
    0.3113   36675080   8. Mautic\CoreBundle\Command\ApplyUpdatesCommand->execute() /Users/lukas.drahy/dev/test/vendor/symfony/console/Command/Command.php:255
  109.5779   37288800   9. Container9uztwgj\appProdProjectContainer->get() /Users/lukas.drahy/dev/test/app/bundles/CoreBundle/Command/ApplyUpdatesCommand.php:246
  109.5780   37288800  10. Container9uztwgj\appProdProjectContainer->load() /Users/lukas.drahy/dev/test/vendor/symfony/dependency-injection/Container.php:304

PHP Fatal error:  require(): Failed opening required '/Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/getMautic_Helper_LanguageService.php' (include_path='/Users/lukas.drahy/dev/test/vendor/phpunit/phpunit-selenium:.:/usr/local/Cellar/php@7.2/7.2.28/share/php@7.2/pear') in /Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/appProdProjectContainer.php on line 5707
PHP Stack trace:
PHP   1. {main}() /Users/lukas.drahy/dev/test/bin/console:0
PHP   2. Symfony\Bundle\FrameworkBundle\Console\Application->run() /Users/lukas.drahy/dev/test/bin/console:41
PHP   3. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:148
PHP   4. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:74
PHP   5. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:255
PHP   6. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:86
PHP   7. Mautic\CoreBundle\Command\ApplyUpdatesCommand->run() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:1001
PHP   8. Mautic\CoreBundle\Command\ApplyUpdatesCommand->execute() /Users/lukas.drahy/dev/test/vendor/symfony/console/Command/Command.php:255
PHP   9. Container9uztwgj\appProdProjectContainer->get() /Users/lukas.drahy/dev/test/app/bundles/CoreBundle/Command/ApplyUpdatesCommand.php:246
PHP  10. Container9uztwgj\appProdProjectContainer->load() /Users/lukas.drahy/dev/test/vendor/symfony/dependency-injection/Container.php:304

Fatal error: require(): Failed opening required '/Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/getMautic_Helper_LanguageService.php' (include_path='/Users/lukas.drahy/dev/test/vendor/phpunit/phpunit-selenium:.:/usr/local/Cellar/php@7.2/7.2.28/share/php@7.2/pear') in /Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/appProdProjectContainer.php on line 5707

Call Stack:
    0.0003     403184   1. {main}() /Users/lukas.drahy/dev/test/bin/console:0
    0.0283    4286320   2. Symfony\Bundle\FrameworkBundle\Console\Application->run() /Users/lukas.drahy/dev/test/bin/console:41
    0.0521    4672824   3. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:148
    0.3098   36609504   4. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:74
    0.3101   36610120   5. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:255
    0.3101   36610120   6. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/lukas.drahy/dev/test/vendor/symfony/framework-bundle/Console/Application.php:86
    0.3111   36674064   7. Mautic\CoreBundle\Command\ApplyUpdatesCommand->run() /Users/lukas.drahy/dev/test/vendor/symfony/console/Application.php:1001
    0.3113   36675080   8. Mautic\CoreBundle\Command\ApplyUpdatesCommand->execute() /Users/lukas.drahy/dev/test/vendor/symfony/console/Command/Command.php:255
  109.5779   37288800   9. Container9uztwgj\appProdProjectContainer->get() /Users/lukas.drahy/dev/test/app/bundles/CoreBundle/Command/ApplyUpdatesCommand.php:246
  109.5780   37288800  10. Container9uztwgj\appProdProjectContainer->load() /Users/lukas.drahy/dev/test/vendor/symfony/dependency-injection/Container.php:304


In appProdProjectContainer.php line 5707:

  Compile Error: require(): Failed opening required '/Users/lukas.drahy/dev/test/var/cache/prod/Container9uztwgj/getMautic_Helper_LanguageService.php' (include_path='/Users/lukas.drahy/dev/test/vendor/phpunit/phpunit-selenium:.:/usr/local/Cellar/php@7.2/7.2.28/share/php@7.2/pear')


mautic:update:apply [--force] [-p|--update-package [UPDATE-PACKAGE]] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command>
```

</details>

**Comment by alanhartless** [comment]
@hluchas Thanks for testing. I apologize for the command issue. I fixed 8466 in order to test the command but obviously didn't actually come back and test nor update the description. 

It seems some more thought is going to have to go into addressing this because of how Symfony nows generates multiple files in the container. When the update command wipes out the cache, it deletes those files, which then the current PHP process can no longer access. I've tried to build a new AppKernel after deleting the cache hoping it would build a new container, get the container, and set it in the current PHP process but that doesn't seem to work. Next I'll try just leaving the original container till the end to see what happens but I fear that may not work either since all Mautic files have not been overwritten with new ones meaning signatures may have changed. 

I'll see if I can figure out a solution.

**Comment by hluchas** [comment]
I noticed then that it's impossible to repeat update workflow in CLI, because is somehow broken. This should be next step for testing and we will need to create separate ticket for this probably

**Comment by alanhartless** [comment]
Lovely; I just trashed this. Tried to rebase but had lots of conflicts. Will have to clean this up tomorrow.

**Comment by alanhartless** [comment]
Ok the command is working now but has to be done in two steps as compared to M2's one. The first run will update the files and wipe out the container. The second run with the `--finish` flag will update translations, update the schema, and do some cleanup.  This is due to Symfony 3's container structure where it builds services in separate files. If the container in PHP's memory tries to load a file that doesn't exist, the whole thing goes kaput.

**Review by dongilbert — APPROVED** [review]
+1 Tested every process for updating. Pushed 1 commit in order to make Travis happy by fixing the tests. Good work @alanhartless!

**Pull request merged**